tests/project/test_task4.py

```diff
@@ -1,6 +1,204 @@
-import pytest
+from GramatykiGrafoweAGH import Graph, Node
+from GramatykiGrafoweAGH.project.task4 import P7
+from GramatykiGrafoweAGH.testing import assert_production_cannot_be_applied
 
 
-@pytest.mark.skip(reason='not implemented yet')
-def test_P7():
-    pass
+def make_P7_left_side_graph():
+    G = Graph()
+
+    level = 2
+
+    E = Node(label='E', x=0, y=0, level=level)
+    iL = Node(label='i', x=-1, y=-1, level=level)
+    iR = Node(label='i', x=1, y=-1, level=level)
+
+    I1L = Node(label='I', x=-2, y=-2, level=level + 1)
+    I2L = Node(label='I', x=-1, y=-2, level=level + 1)
+
+    I1R = Node(label='I', x=1, y=-2, level=level + 1)
+    I2R = Node(label='I', x=2, y=-2, level=level + 1)
+
+    E1L = Node(label='E', x=0, y=-3, level=level + 1)
+    E2L = Node(label='E', x=0, y=-4, level=level + 1)
+    E3L = Node(label='E', x=0, y=-5, level=level + 1)
+
+    E1R = Node(label='E', x=0, y=-3, level=level + 1)
+    E2R = Node(label='E', x=0, y=-4, level=level + 1)
+    E3R = Node(label='E', x=0, y=-5, level=level + 1)
+
+    G.add_nodes([E, iL, iR, I1L, I2L, I1R, I2R, E1L, E2L, E3L, E1R, E2R, E3R])
+
+    G.add_edges([
+        (E, iL), (E, iR), (iL, I1L), (iL, I2L), (iR, I1R), (iR, I2R),
+        (I1L, E1L), (I1L, E2L), (I2L, E2L), (I2L, E3L),
+        (I1R, E1R), (I1R, E2R), (I2R, E2R), (I2R, E3R),
+        (E1L, E2L), (E2L, E3L), (E1R, E2R), (E2R, E3R)
+    ])
+
+    return G
+
+
+def make_P7_left_side_graph_not_middle():
+    G = Graph()
+
+    level = 2
+
+    E = Node(label='E', x=0, y=0, level=level)
+    iL = Node(label='i', x=-1, y=-1, level=level)
+    iR = Node(label='i', x=1, y=-1, level=level)
+
+    I1L = Node(label='I', x=-2, y=-2, level=level + 1)
+    I2L = Node(label='I', x=-1, y=-2, level=level + 1)
+
+    I1R = Node(label='I', x=1, y=-2, level=level + 1)
+    I2R = Node(label='I', x=2, y=-2, level=level + 1)
+
+    E1L = Node(label='E', x=0, y=-3, level=level + 1)
+    E2L = Node(label='E', x=0, y=-4.2, level=level + 1)
```

Comment:
Instead of duplicating the whole implementation of `make_P7_left_side_graph`, it probably would be better to call `make_P7_left_side_graph` and then find nodes `E2L` and `E2R` and simply replace them in the specific test.

Reply:
That is a good suggestion, applied!